### DISCUSSION TITLE
How to select a specific repeating HL7 2.x PID-3 segment based on value of PID-3.4.1?

### DISCUSSION THREAD

**Discussion by gh69** [nextgenhealthcare/connect#5966] [Q&A] [answered]
Hello,

I'm brand new to Mirth Connect and I'm running version 4.4.0.  I've created a channel that takes HL7 input from a file, does some minor filtering/transforms, and then spits the output into another file.  What I'm having trouble figuring out is how to deal with repeating segments.  Specifically, I've got inbound messages that have multiple repeats within PID-3 and I'm trying to remove all repeats except one, based solely on the value in PID-3.4.1 (that value being "NEW_MRN").  Looking through the forum messages I see that similar questions have been asked and in trying to overwrite the PID-3 with the value of whichever repeated PID-3 contains a PID-3.4.1 value of "NEW_MRN", this is the what I've been able to cobble together but it doesn't seem to work.  Any experts out there who might be able to guide me on this?

Javascript transformer on the Destination:
```
msg.PID['PID.3'] = msg.PID['PID.3'].(function(){return (new XML(this))['PID.3.4.1'] == 'NEW_MRN'}());
```

INBOUND MESSAGE:
```
MSH|^~\&||SENDFAC|||20230929075609+0000||ADT^A08^ADT_A01|Q0T0X0||2.8.2|||AL|NE||8859/1
EVN||20230929075609+0000
PID|1||A111^^^OLD_MRN_A&0.00.000.0.000000.0.0000.000&ISO^TEXT^TEXT^19700101000000+0000~B222^^^OLD_MRN_B&1.11.111.1.111111.1.1111.111&ISO^TEXT^TEXT^19700101000000+0000~777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000||LAST^FIRST||19700101|M||||||||||9000
PV1|1|Outpatient^Outpatient|LOC^^^LOC^^AMB^LOC||||12345^JOHNSON^DOC^^NP|12345^JOHNSON^DOC^^NP||||||||||||||||||||||||||||01^01 Home or Self Care|||||Discharged^Discharged|||19700101000000+0000|19700101000000+0000
```

OUTBOUND MESSAGE:
```
MSH|^~\&||SENDFAC|||20230929075609+0000||ADT^A08^ADT_A01|Q0T0X0||2.8.2|||AL|NE||8859/1
EVN||20230929075609+0000
PID|1||A111^^^OLD_MRN_A&0.00.000.0.000000.0.0000.000&ISO^TEXT^TEXT^19700101000000+0000~B222^^^OLD_MRN_B&1.11.111.1.111111.1.1111.111&ISO^TEXT^TEXT^19700101000000+0000~777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000||LAST^FIRST||19700101|M||||||||||9000
PV1|1|Outpatient^Outpatient|LOC^^^LOC^^AMB^LOC||||12345^JOHNSON^DOC^^NP|12345^JOHNSON^DOC^^NP||||||||||||||||||||||||||||01^01 Home or Self Care|||||Discharged^Discharged|||19700101000000+0000|19700101000000+0000
```

EXPECTED OUTBOUND MESSAGE:
```
MSH|^~\&||SENDFAC|||20230929075609+0000||ADT^A08^ADT_A01|Q0T0X0||2.8.2|||AL|NE||8859/1
EVN||20230929075609+0000
PID|1||777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000||LAST^FIRST||19700101|M||||||||||9000
PV1|1|Outpatient^Outpatient|LOC^^^LOC^^AMB^LOC||||12345^JOHNSON^DOC^^NP|12345^JOHNSON^DOC^^NP||||||||||||||||||||||||||||01^01 Home or Self Care|||||Discharged^Discharged|||19700101000000+0000|19700101000000+0000
```

When I run this through, the output file contains the entire PID-3 segment from the input file.

**Comment by pacmano1** [comment]
Please use code blocks.  Read https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

```
PID|1||A111^^^OLD_MRN_A&0.00.000.0.000000.0.0000.000&ISO^TEXT^TEXT^19700101000000+0000B222^^^OLD_MRN_B&1.11.111.1.111111.1.1111.111&ISO^TEXT^TEXT^19700101000000+0000777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000
```

```JS
msg.PID['PID.3'] = msg.PID['PID.3'].(function(){return (new XML(this))['PID.3.4.1'] == 'NEW_MRN'}());
```

Please post a before and expected "after" message.   Please post a complete message since if someone is messing with this in their engine they need a complete message anyway.

**Reply by gh69** [reply]
Thanks.  I've updated the original text to comply with this format.

**Reply by pacmano1** [reply]
assuming you are not searching for 'NEW_MRN'

```JS
msg['PID']['PID.3'] = msg['PID']['PID.3'][2]
```

If you are searching, can just use a for loop to get the index.

Before:
```
MSH|^~\&||SENDFAC|||20230929075609+0000||ADT^A08^ADT_A01|Q0T0X0||2.8.2|||AL|NE||8859/1
EVN||20230929075609+0000
PID|1||A111^^^OLD_MRN_A&0.00.000.0.000000.0.0000.000&ISO^TEXT^TEXT^19700101000000+0000~B222^^^OLD_MRN_B&1.11.111.1.111111.1.1111.111&ISO^TEXT^TEXT^19700101000000+0000~777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000||LAST^FIRST||19700101|M||||||||||9000
PV1|1|Outpatient^Outpatient|LOC^^^LOC^^AMB^LOC||||12345^JOHNSON^DOC^^NP|12345^JOHNSON^DOC^^NP||||||||||||||||||||||||||||01^01 Home or Self Care|||||Discharged^Discharged|||19700101000000+0000|19700101000000+0000
```
Code above in source transformer, AFTER:
```
MSH|^~\&||SENDFAC|||20230929075609+0000||ADT^A08^ADT_A01|Q0T0X0||2.8.2|||AL|NE||8859/1
EVN||20230929075609+0000
PID|1||777^^^NEW_MRN&0.00.000.0.000000.0.0000.000&ISO^MRN^TEXT^19700101000000+0000||LAST^FIRST||19700101|M||||||||||9000
PV1|1|Outpatient^Outpatient|LOC^^^LOC^^AMB^LOC||||12345^JOHNSON^DOC^^NP|12345^JOHNSON^DOC^^NP||||||||||||||||||||||||||||01^01 Home or Self Care|||||Discharged^Discharged|||19700101000000+0000|19700101000000+0000
```

**Reply by gh69** [reply]
Yes, I am searching for NEW_MRN, which can be in any repeat, sometimes it's in [2] but not always...but I think I see what you're getting at.  Let me see if I can figure out how to find the index of the repeated PID-3 that contains NEW_MRN and then pass that index to the code you provided.  Thanks :D

**Comment by jonbartels** [comment]
https://github.com/search?q=repo%3Anextgenhealthcare%2Fconnect-examples%20xfilter&type=code would the xFilter example work for this use case?

**Reply by pacmano1** [reply]
It would.

**Reply by gh69** [reply]
I'll check this out as well.  Thanks Jonb.

**Comment by gh69** [comment]
Based on Pacman's suggestion, this is what I came up with.  I feel like it could be tighter but it does work:

```
for each (repeat in msg['PID']['PID.3'])
    {
     if (repeat['PID.3.4']['PID.3.4.1'].toString()=="NEW_MRN")
        {
         msg['PID']['PID.3'] = msg['PID']['PID.3'][index];        
         break;
        }
     index++;
    }
```

**Reply by pacmano1** [reply, marked as answer]
Make sure you add the xFilter function to code templates and it is accessible to your channel (or just paste it into your transformer for now)

```JS
var mysegment = xFilter(msg.PID['PID.3'], function(pid) {
    return pid['PID.3.4']['PID.3.4.1'] == 'NEW_MRN';
}, 0, 1);

msg['PID']['PID.3'] = mysegment
```

**Reply by tonygermano** [reply]
If you want to shorten it further and make it more like your original code:
```js
msg['PID']['PID.3'] = xFilter(msg.PID['PID.3'], pid3 => pid3['PID.3.4']['PID.3.4.1'] == 'NEW_MRN')
```

Also, I think your original code wasn't working because you had
```js
msg.PID['PID.3'] = msg.PID['PID.3'].(function(){return (new XML(this))['PID.3.4.1'] == 'NEW_MRN'}());
// instead of
msg.PID['PID.3'] = msg.PID['PID.3'].(function(){return (new XML(this))['PID.3.4']['PID.3.4.1'] == 'NEW_MRN'}());
```
